hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageSessionFactoryImpl.java

```diff
@@ -201,10 +192,17 @@ public <T> CompletionStage<T> withStatelessSession(String tenantId, Function<Sta
 		}
 		else {
 			LOG.debugf( "No existing open Stage.StatelessSession was found in the current Vert.x context for current tenant '%s': opening a new instance", tenantId );
-			return withSession( openStatelessSession( tenantId), work, key );
+			return oneOffExecution( v -> withSession( openStatelessSession( tenantId ), work, contextKeyForStatelessSession ) );
 		}
 	}
 
+	private <T> CompletionStage<T> oneOffExecution(Function<Void, CompletionStage<T>> fun) {
+		OneOffDelegatingExecutor oneOffExecutor = new OneOffDelegatingExecutor( context );
+		final CompletionStage<T> withSessionStage = voidFuture().thenComposeAsync( fun, oneOffExecutor );
+		oneOffExecutor.runHeldTasks();
+		return withSessionStage;
```

Comment:
Interesting, but if this is the final pattern then the `OneOffDelegatingExecutor` should not be necessary?
The function is completely defined at this point, so there are no additional stages that still need the Session as downstream events: you could run it with `thenComposeAsync` directly.